gel/_internal/_qbmodel/_abstract/_primitive.py

```diff
@@ -403,26 +430,206 @@ def get_py_type_scalar_match_rank(
 _PT_co = TypeVar("_PT_co", bound=PyConstType, covariant=True)
 
 
-def get_literal_for_value(t: type[PyConstType], v: PyConstType) -> _qb.Literal:
+def get_literal_for_value(
+    v: Any,
+) -> _qb.Literal:
+    for t, ltype in _py_type_to_literal.items():
+        if isinstance(v, t):
+            return ltype(val=v)  # type: ignore [call-arg]
+
+    raise NotImplementedError(f"cannot convert Python value to Literal: {v!r}")
+
+
+def get_literal_for_scalar(
+    t: type[PyTypeScalar[_PT_co]],
+    v: Any,
+) -> _qb.Literal | _qb.CastOp:
     if not isinstance(v, t):
-        raise ValueError(
-            f"get_literal_for_value: {v!r} is not an instance of {t}"
-        )
-    ltype = _py_type_to_literal.get(t)
-    if ltype is None:
-        raise NotImplementedError(f"unsupported Python raw value type: {t}")
-    else:
-        if t is builtins.bool:
-            v = builtins.bool(v)
+        v = t(v)
+    ltype = _py_type_to_literal.get(t.type)  # type: ignore [arg-type]
+    if ltype is not None:
         return ltype(val=v)  # type: ignore [call-arg]
+    else:
+        return _qb.CastOp(
+            expr=_qb.StringLiteral(val=str(v)),
+            type_=t.__gel_reflection__.name,
+        )
 
 
-class PyTypeScalar(_typing_parametric.SingleParametricType[_PT_co]):
+class PyTypeScalar(
+    _typing_parametric.SingleParametricType[_PT_co],
+    BaseScalar,
+):
     if TYPE_CHECKING:
 
-        def __init__(self, val: Any) -> None: ...
         def __set__(self, obj: Any, value: _PT_co) -> None: ...  # type: ignore [misc]
 
-    def __edgeql_literal__(self) -> _qb.Literal:
-        pytype = type(self).type
-        return get_literal_for_value(pytype, self)  # type: ignore [arg-type]
+    def __edgeql_literal__(self) -> _qb.Literal | _qb.CastOp:
+        return get_literal_for_scalar(type(self), self)
+
+
+UUIDFieldsTuple = TypeAliasType(
+    "UUIDFieldsTuple", tuple[int, int, int, int, int, int]
+)
+
+
+class UUIDImpl(uuid.UUID):
```

Comment:
why do we need these wrappers?

Reply:
To allow downcasting, i.e `std.uuid(uuid.UUID(...))`, because standard `uuid.UUID` class does not support this.